lib/modules/datasource/pypi/index.ts

```diff
@@ -43,45 +46,51 @@ export class PypiDatasource extends Datasource {
     packageName,
     registryUrl,
   }: GetReleasesConfig): Promise<ReleaseResult | null> {
-    let dependency: ReleaseResult | null = null;
     // TODO: null check (#22198)
-    const hostUrl = ensureTrailingSlash(
-      registryUrl!.replace('https://pypi.org/simple', 'https://pypi.org/pypi'),
-    );
+    const hostUrl = ensureTrailingSlash(registryUrl!);
+    const simpleHostUrl = hostUrl.replace(jsonPyPiUrl, simplePyPiUrl);
+    const pypiJsonHostUrl = hostUrl.replace(simplePyPiUrl, jsonPyPiUrl);
     const normalizedLookupName = normalizePythonDepName(packageName);
 
-    // not all simple indexes use this identifier, but most do
-    if (hostUrl.endsWith('/simple/') || hostUrl.endsWith('/+simple/')) {
+    const simpleDependencies = await this.getResultsViaSimple(
+      normalizedLookupName,
+      simpleHostUrl,
+    ).catch((err) => {
       logger.trace(
-        { packageName, hostUrl },
-        'Looking up pypi simple dependency',
+        { packageName, hostUrl: simpleHostUrl },
+        'Simple api not found. Looking up pypijson api as fallback.',
       );
-      dependency = await this.getSimpleDependency(
-        normalizedLookupName,
-        hostUrl,
+      return null;
+    });
+    logger.trace(
+      { packageName, hostUrl: pypiJsonHostUrl },
+      'Querying json api for metadata',
+    );
+    const pypiJsonDependencies = await this.getResultsViaPyPiJson(
+      normalizedLookupName,
+      pypiJsonHostUrl,
+    ).catch((err) => {
+      if (simpleDependencies === null) {
+        throw err;
+      }
+      logger.trace(
+        { packageName, hostUrl },
+        'Json api lookup failed but got simple results.',
       );
-    } else {
-      logger.trace({ packageName, hostUrl }, 'Looking up pypi api dependency');
-      try {
-        // we need to resolve early here so we can catch any 404s and fallback to a simple lookup
-        dependency = await this.getDependency(normalizedLookupName, hostUrl);
-      } catch (err) {
-        if (err.statusCode !== 404) {
-          throw err;
-        }
-
-        // error contacting json-style api -- attempt to fallback to a simple-style api
-        logger.trace(
-          { packageName, hostUrl },
-          'Looking up pypi simple dependency via fallback',
-        );
-        dependency = await this.getSimpleDependency(
-          normalizedLookupName,
-          hostUrl,
-        );
+      return null;
```

Comment:
Won't this return `null` incorrectly? i.e. simple succeeded, API failed, you return `null`

Reply:
I think the return here ends up being the value of `pypiJsonDependencies`.